### DISCUSSION TITLE
System Testing Plan for Worker

### DISCUSSION THREAD

**Discussion by adrianmichalke** [nebulastream/nebulastream#20] [Ideas]
**TLDR.** Most systems heavily rely on system level tests on the user facing level, e.g., SQL. Test like this in NES are executed inefficiently and difficult to write. 

## How other Systems are Tested:
- **[DuckDB](https://github.com/duckdb):**
	- Example CI Run
	- [Duckdb testing philosophy](https://duckdb.org/docs/dev/sqllogictest/overview.html#:~:text=We%20run%20a%20batch%20of,issues%20found%20through%20fuzzing%20DuckDB.)
		- > When testing DuckDB, we aim to route all the tests through SQL. We try to avoid testing components individually because that makes those components more difficult to change later on.  As such, almost all of our tests can (and should) be expressed in pure SQL.
- **[FoundationDB](https://github.com/apple/foundationdb):**
	- Example CI Run (NA hidden in AWS)
	- https://github.com/apple/foundationdb/tree/main/tests
	- Tests are defined in [declarative self-contained files](https://github.com/apple/foundationdb/blob/main/tests/slow/LongRunning.toml) testing E2E
	- Determininstic Testing framework for distributed tests
	    - ["Testing Distributed Systems w/ Deterministic Simulation" YT](https://www.youtube.com/watch?v=4fFDFbi3toc&ab_channel=StrangeLoopConferenc)
	- Tests grouped into: fast, slow, rare, restarting, ...
- **[Clickhouse](https://github.com/ClickHouse/ClickHouse/):**
	- [Example CI Run](https://github.com/ClickHouse/ClickHouse/actions/runs/9417818222)
	- https://clickhouse.com/docs/ru/development/tests
	    - > Functional tests are the most simple and convenient to use. Most of ClickHouse features can be tested with functional tests and they are mandatory to use for every change in ClickHouse code that can be tested that way. Each functional test sends one or multiple queries to the running ClickHouse server and compares the result with reference.
	      >
	      >  It’s not necessary to have unit tests if the code is already covered by functional tests (and functional tests are usually much more simple to use).
- **[Flink](https://github.com/apache/flink):**
	- [Example CI Run](https://github.com/apache/flink/actions/runs/9462873581)
	- E2E Tests (divided into two parallel running groups each ~2h)
	   - E2E test are written as [bash scripts](https://github.com/apache/flink/blob/526f9b034763fd022a52fe84b2c3227c59a78df1/flink-end-to-end-tests/test-scripts/test_tpcds.sh)
	- Unit Tests (core, python, table, connect, tests, misc)
- **[Spark](https://github.com/apache/spark):**
	- [Example CI Run](https://github.com/panbingkun/spark/actions/runs/9411155017)
	- E2E Tests in SQL divided into: slow tests, extended tests, other tests
	   - [Tests](https://github.com/panbingkun/spark/blob/c4f720dfb41919dade7002b49462b3ff6b91eb22/sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala) written in scala issuing SQL queries
	- Unit tests for each module
- **[Materialize](https://github.com/MaterializeInc/materialize):**
	- [Example CI Run](https://buildkite.com/materialize/nightly/builds/8021#_)
	- [Materialize Testing Philosophy](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-testing.md)
         - > There are broadly three test suites:            
            > - The Rust unit/integration test suite, which is run by cargo test. These live alongside the code they test in a mod tests { ... } block, or in the tests/ directory within a crate.
              > - The data-driven system test suite in the top-level test/directory. These consist of text files in various DSLs (sqllogictest, testdrive, pgtest) that essentially specify SQL commands to run and their expected output.
              > - The long-running performance and stability test suite. This test suite has yet to be automated. At the moment it consists of engineers manually running the demo in test/chbench.

             > The unit/integration and system test suites are run on every PR and can easily be run locally. The goal is for these test suites to be quite fast, ideally under 10m, while still catching 95% of bugs.

              > The long-running tests will simulate actual production runs of Materialize on gigabytes to terabytes of data over several hours to days, and will therefore report their results asynchronously (perhaps nightly or weekly).

Takeaways:
- Most of the systems heavily rely on testing on the SQL/User facing level.
- These tests are easy to write in standardized way.
- Some apply a hierarchy of testing (smoke tests --> fast tests --> slow test --> long running tests)

## System Tests

The worker should have clear interfaces after the first milestone, e.g.: register query, start query, stop query.
To make our worker bulletproof our tests should test on these interfaces extensively and we should make writing these tests easy and its execution efficient.

**Despite being slower, system level tests have a lot of advantages:**
1. They are independent from internal components, makes it easier to change these components later. When changing internals of the worker the all system level test should still hold.
2. They test how all components of the worker work together.
3. One system level test can make multiple unit test obsolete. E.g. [adding an operator](https://docs.nebula.stream/docs/dev/guide/operators/) leads to 8 test on different level of the lowering pipeline. When using a system test instead we can track if we still cover all our code paths by enforcing PR test coverage.
4. They are easier to write, in the best case only a query.

**System level tests in NES:**
E.g. see: [Multiple Windows Test](https://github.com/nebulastream/nebulastream/blob/master/nes-coordinator/tests/Integration/Windowing/MultipleWindowsTest.cpp)

Many of our test on this level are very slow, e.g. many integration tests, start coordinator and worker in each test instance. Also they require a lot of code to write to just issue queries.

**Comment by adrianmichalke** [comment]
### Idea

**How to make system level tests faster:**
We should bundle test:
- Start worker once, run all tests, stop worker
- In theory, test queries could run in parallel on one worker, but we have to see

**How to make system level tests easier to write:**
- Tests are self-contained files similar to duckDB/FoundationDB/MLIR:
```SQL
# name: test/sql/projection/test_simple_projection.test
# group [projection]

# enable query verification
statement ok
PRAGMA enable_verification

# create table
statement ok
CREATE TABLE a (i INTEGER, j INTEGER);

# insertion: 1 affected row
statement ok
INSERT INTO a VALUES (42, 84);

query II
SELECT * FROM a;
----
42	84
```
- Test file contains i) query written in CQL/End-user-facing language, ii) the input stream file, and  iii) expected result.
- If the expected result is not easy to derive, we might test the same query against Flink (in the future also with other techniques, e.g., metamorphic testing) 
- Running multiple of these tests at the same time on one worker should be possible, but we have to see.

**Reply by keyseven123** [reply]
I really like the idea of having such files. Later, we can replace the syntax with our own, if we wish to do so. 

> Start worker once, run all tests, stop worker
> In theory, test queries could run in parallel on one worker, but we have to see

One potential problem I see here is that if a query fails, it might create an unwanted state that will lead to other queries failing. Maybe we can track what queries fail and let them then run in an isolated environment (worker runs one query and gets shutdown after each query).

**Reply by alepping** [reply]
I agree. Such testfiles enable a coherent way to test whether queries run successfully.
In my opinion, we should decide what the lowest granularity for our tests are.
Potentially, this level could be querying on the worker level. If this takes too long (grpc etc.) it might be the 'NodeEngine' level.
In the long run, we should also think about how to test distributed setups, control flow, statistics, etc.(parts of our system that are difficult to test with user queries (data processing) only (given that queries do not allow to test for these things specifically)).

**Reply by adrianmichalke** [reply]
> In my opinion, we should decide what the lowest granularity for our tests are.
> Potentially, this level could be querying on the worker level. If this takes too long (grpc etc.) it might be the 'NodeEngine' level.

100%, I think the first task for this epic will be to check how fast we can get these test on the worker level. If they are to slow we should consider to test on a lower level.

**Reply by adrianmichalke** [reply]
_/closed, covered in issue https://github.com/nebulastream/nebulastream-public/issues/31_

**Comment by PhilippGrulich** [comment]
I like this way of testing. I used something similar in nautilus and it is quite easy. 
I think this would be really nice if we have a way to write such test files with a text representation of the query.

**Comment by ls-1801** [comment]
I like it, maybe this way we can create a catalog of _must work at all_ times queries, that we can use during our stability tests.

However I would exclude these tests from the testcoverage, as they very effective at increasing coverage without testing failure or edge cases.

Do you want to generate the data or have the data prepared? 

I don't want to overcomplicate this, because I think this is 100% required for milestone 1, so even the simplest form of these tests is already very valuable.

**Reply by adrianmichalke** [reply]
Im unsure about the exclusion from the testcoverage because these test go through code paths in the same way as unit/integration/etc tests do. 

Also, I see not problem in writing system tests that we expect to fail. But as we already discussed offline we can see how it turns out and postpone the decision.

For now I would like to have data prepared. For small queries like expressions we can have the query, the input and expected result in the same file.

**Comment by zeuchste** [comment]
### Testing Rules
Currently, we do have a lot of disabled tests, what is our approach here, do we also want to activate or remove all of them? Does a deactivated test not indicate that there is a not used feature and thus we should remove it?

**Reply by keyseven123** [reply]
Yes, I agree. If the majority of tests for a component is disabled, we should remove this component for now and if needed reimplement it later.

**Reply by adrianmichalke** [reply]
Yes, I agree too. I can create an issue to track tests. Regarding untested components we might want to use the testcoverage reports

**Comment by zeuchste** [comment]
### Regression tests for performance

I would really appreciate if you plan for testing also includes a continuous running framework to identify performance regressions, for example using a nightly build and report performance degeneration to make sure we do not hamper the performance significantly with what we do

**Reply by adrianmichalke** [reply]
Yes, I think that would be very cool and after we have the test framework foundation we can implement that quite easily and define queries we want to performance test.

I kind of like the regression test reports that clickhouse produces: https://s3.amazonaws.com/clickhouse-test-reports/0/afba7bb4edbdc1423c6b8dde1eca79fedb777fdc/performance_comparison_[4_4]/report.html

Mutable has a sweet dashboard for that publicly available: https://cb.mutable.uni-saarland.de/dashboard

**Reply by fwc** [reply]
https://conbench.github.io/conbench/ also seems interesting at first glance
